tac: add support for --regex option to tac

jfinkels opened the pull request:
Add support for `tac --regex`, where the line separator is interpreted
as a regular expression.

sylvestre commented on Cargo.lock:
```diff
@@ -2948,6 +2948,7 @@ version = "0.0.7"
 dependencies = [
```
yeah

sylvestre commented on tests/by-util/test_tac.rs:
```diff
@@ -205,3 +205,21 @@ fn test_null_separator() {
+}
+
+#[test]
+fn test_regex_before() {
```
would it be possible to add some more complex regex to testing?

jfinkels replied:
Sure, I can add some more.

sylvestre replied:
ping?

miDeb commented on src/uu/tac/src/tac.rs:
```diff
@@ -81,6 +81,54 @@ pub fn uu_app() -> App<'static, 'static> {
+    // of one match and the start of the next. Otherwise, a line is
+    // between the end of one match and the end of the next.
+    let mut endpoints: Vec<usize> = pattern
+        .find_iter(data)
```
This is not 100% correct because the regex searches from the front while it should be searching from the back. I'm not sure how easy it is to fix, because the `regex` crate does not provide `find_reverse` or similar.
Here's an example where it matters: 
`printf "axxxb" | tac --separator="xx"` should print `baxxx`, but if you enable `--regex` it prints `xbaxx` instead.

miDeb replied:
We basically want to search for the separator in `data[last_separator_index - 1..last_separator_index]`, then, if there's no match, search in `data[last_separator_index - 2..last_separator_index]`, and so on. This sounds like it would be very expensive, but maybe a possible solution? FWIW, this is what GNU seems to be doing.

jfinkels replied:
Good catch. I'm probably not the right person to implement a complete `find_reverse()`, but I might be able to produce a naive version that covers basic use cases like the one you just described. I will update this pull request with that change.

jfinkels commented:
Hm, the test case I added is not even correct. I'm not sure why I didn't notice that. Here's the real behavior of GNU `tac` version 8.28:
```
$ printf "a:b::c:::d::::" | tac -r -s ":+" && echo
:::d:::c::b:a:
```
Here is a slightly simpler example that better demonstrates the behavior.
```
$ printf "axyz" | tac -r -s "[xyz]+" && echo
zyax
```
I think this matches what @miDeb suggested above, but I just want to double check my understanding.

* Scanning from right to left, the character "z" is a match, so that is taken as a line ending. Same with "y". The slice starting after "y" and ending after "z" is the empty string followed by "z". That's the first line printed.
* Next, we continue scanning to the left from "y" and we find "x". The slice starting after "x" and ending after "y" is the empty string followed by "y". That's the second line printed.
* Next, we continue scanning to the left from "x" and find "a". The character "a" is not a match, so we continue to the left. But we are at the left end of the text, so the slice starting at the left end of the text and ending after "x" is "a" followed by "x". That's the third line printed.

miDeb commented:
@jfinkels yes, this is correct. I'd like to point out that the trailing `+` in the regex has no effect, because anything it could match was already matched as a separator before. For this reason your first example has the same output as `printf "a:b::c:::d::::" | tac -s ":" && echo`.

jfinkels commented:
I've added a couple more tests and updated the implementation to match the description of how `--regex` is supposed to work based on the comments above.

jfinkels commented:
Clippy is complaining about a line in an unrelated file.

miDeb commented:
Thanks a lot!
